type-c-service/src/wrapper/cfu.rs

```diff
@@ -0,0 +1,281 @@
+    InProgress(u8),
+    /// Firmware update has failed and the device is in an unknown state
+    Recovery,
+}
+
+impl FwUpdateState {
+    /// Check if the firmware update is in progress
+    pub fn in_progress(&self) -> bool {
+        matches!(self, FwUpdateState::InProgress(_) | FwUpdateState::Recovery)
+    }
+}
+
+impl<const N: usize, C: Controller, V: FwVersionValidator> ControllerWrapper<'_, N, C, V> {
+    /// Process a GetFwVersion command
+    async fn process_get_fw_version(&self, target: &mut C) -> InternalResponseData {
+        let version = match target.get_active_fw_version().await {
+            Ok(v) => v,
+            Err(Error::Pd(e)) => {
+                error!("Failed to get active firmware version: {:?}", e);
+                return InternalResponseData::ComponentBusy;
+            }
+            Err(Error::Bus(_)) => {
+                error!("Failed to get active firmware version, bus error");
+                return InternalResponseData::ComponentBusy;
+            }
+        };
+
+        let dev_inf = FwVerComponentInfo::new(FwVersion::new(version), self.cfu_device.component_id());
+        let comp_info: [FwVerComponentInfo; MAX_CMPT_COUNT] = [dev_inf; MAX_CMPT_COUNT];
+        InternalResponseData::FwVersionResponse(GetFwVersionResponse {
+            header: GetFwVersionResponseHeader::new(1, GetFwVerRespHeaderByte3::NoSpecialFlags),
+            component_info: comp_info,
+        })
+    }
+
+    /// Process a GiveOffer command
+    async fn process_give_offer(&self, target: &mut C, offer: &FwUpdateOffer) -> Option<InternalResponseData> {
+        if offer.component_info.component_id != self.cfu_device.component_id() {
+            return None;
+        }
+
+        let version = match target.get_active_fw_version().await {
+            Ok(v) => v,
+            Err(Error::Pd(e)) => {
+                error!("Failed to get active firmware version: {:?}", e);
+                return Some(InternalResponseData::ComponentBusy);
+            }
+            Err(Error::Bus(_)) => {
+                error!("Failed to get active firmware version, bus error");
+                return Some(InternalResponseData::ComponentBusy);
+            }
+        };
+
+        if !self.fw_version_validator.validate(FwVersion::new(version), offer) {
+            debug!("Version is not acceptable, rejecting offer");
+            return None;
+        }
+
+        debug!("Accepting offer");
+        Some(InternalResponseData::OfferResponse(FwUpdateOfferResponse::new_accept(
+            HostToken::Driver,
+        )))
+    }
+
+    /// Process a GiveContent command
+    async fn process_give_content(
+        &self,
+        controller: &mut C,
+        state: &mut InternalState,
+        content: &FwUpdateContentCommand,
+    ) -> Option<InternalResponseData> {
+        let data = &content.data[0..content.header.data_length as usize];
+        if content.header.flags & FW_UPDATE_FLAG_FIRST_BLOCK != 0 {
```

Comment:
no special processing needed after the successful write of LAST_BLOCK ?  When could the controller validate the bits when it has everything?